research/conditional/utils/model_utils.py

```diff
@@ -129,6 +134,86 @@ def calculate_llm_loss(
     model: torch.nn.Module,
     mixed_precision: bool,
     vocab_size: int,
+    n_blanks: int = 0,
+    blank_id: Optional[int] = 50257,
+):
+    input_tokens = batch.input_ids
+    gt_tokens = batch.target_ids
+    mask = batch.should_calculate_loss
+
+    with torch.autocast(
+        device_type="cuda", enabled=mixed_precision, dtype=torch.float16
+    ):
+        model_output = model(input_tokens)
+
+    # move the gt tokens and mask to the same device as the model output - they should be on the same device for loss calculation
+    gt_tokens = gt_tokens.to(model_output.device)
+    mask = mask.to(model_output.device)
+
+    mask_loss = F.cross_entropy(
+        model_output.reshape(-1, vocab_size),
+        gt_tokens.reshape(-1).long(),
+        reduction="none",
+    )
+    mask_loss = mask_loss[mask.reshape(-1) == 1]
+    loss = mask_loss.mean()
+
+    blanks_losses = {}
+    if n_blanks > 0:
+        with torch.no_grad():
+            is_blank = input_tokens.eq(blank_id)
+            # is_blank.shape = (B, S)
+            # TODO: test if this does what we want
+            blank_start = (
+                (
+                    F.conv1d(
+                        is_blank[:, None, :].float(),
+                        torch.Tensor([-1.0, 1.0, 0.0], device=is_blank.device).reshape(
+                            1, 1, -1
+                        ),
+                        padding="same",
+                    )
+                    == 1
+                )
+                .float()
+                .squeeze_(1)
+            )
+
+            # TODO: test this as well
+            # print(input_tokens.shape)
+            # print(blank_start.shape)
+            nth_blank_mask = F.pad(blank_start[:, 1:], (0, 1), "constant", 0)
+            # print(nth_blank_mask.shape)
+            for i in range(n_blanks + 1):
+                if i == 0:
+                    assert not input_tokens[nth_blank_mask == 1].eq(blank_id).any()
+                else:
+                    assert input_tokens[nth_blank_mask == 1].eq(blank_id).all()
+                nth_blank_loss = (nth_blank_mask.reshape(-1) * mask_loss).mean()
+                blanks_losses[f"blank_{i}_loss"] = nth_blank_loss
```

Comment:
adding a diff losses would be great. In particular, for every (positive) blank i, compute diff(i, 0) and diff(i, i-1).

Reply:
not necessary for MVP, maybe, but it'll make it all easier to analyze